### PR TITLE
Fix UTF-8 handling

### PR DESCRIPTION
**Pull request by ihrwein**
UTF-8 wasn't handled well: there were cases, when multibyte characters were splitted resulting a run time panic.

This PR intends to fix this behavior.

### CODE REVIEW

**Comment by ihrwein** [comment]
I don't see the reason how the coveralls could fall... This PR contains mostly fixes, new unit tests and a refactor (which eliminates some code).

**Comment by bkil-syslogng** [comment]
Just a guess. Maybe coverage is only recomputed on master, while total code lines is summed as a divider.

File: src/matcher/suffix_array/test.rs [hunk 1]
```diff
@@ -146,3 +146,18 @@ fn test_given_suffix_array_when_literals_are_inserted_then_it_can_find_the_strin
 
     assert_eq!("app", root.longest_common_prefix("app42").unwrap().literal());
 }
+
+#[test]
+fn test_given_parser_when_it_receives_utf_8_strings_then_it_does_not_panic() {
+    let string_parser = "%{GREEDY}¡%{GREEDY}";
```
Comment (lbudai) [inline src/matcher/suffix_array/test.rs]: so this is a pattern (and not parser)

**Comment by lbudai** [comment]
:+1:

**Comment by MrAnno** [comment]
:+1:

**Comment by ihrwein** [comment]
Autosquash done, PR is ready for merge after CI is green.

**Pull request merged**